src/query.rs

```diff
@@ -8,113 +8,189 @@ use http::{
 use reqwest::Client;
 use tokio::time::sleep;
 
+use crate::row::SnowflakeColumnType;
+use crate::SnowflakeSession;
 use crate::{chunk::download_chunk, Error, Result, SnowflakeRow};
 
 pub(super) const SESSION_EXPIRED: &str = "390112";
 pub(super) const QUERY_IN_PROGRESS_ASYNC_CODE: &str = "333334";
+const DEFAULT_TIMEOUT_SECONDS: u64 = 300;
+
+pub struct QueryExecutor {
+    http: Client,
+    qrmk: String,
+    chunks: Vec<RawQueryResponseChunk>,
+    chunk_headers: HeaderMap,
+    column_types: Arc<HashMap<String, (usize, SnowflakeColumnType)>>,
+    row_set: Option<Vec<Vec<Option<String>>>>,
+}
 
-pub(super) async fn query<Q: Into<QueryRequest>>(
-    http: &Client,
-    account: &str,
-    request: Q,
-    session_token: &str,
-    timeout: Duration,
-) -> Result<Vec<SnowflakeRow>> {
-    let request_id = uuid::Uuid::new_v4();
-    let url = format!(
-        r"https://{account}.snowflakecomputing.com/queries/v1/query-request?requestId={request_id}"
-    );
-
-    let request: QueryRequest = request.into();
-    let response = http
-        .post(url)
-        .header(ACCEPT, "application/snowflake")
-        .header(
-            AUTHORIZATION,
-            format!(r#"Snowflake Token="{}""#, session_token),
-        )
-        .json(&request)
-        .send()
-        .await?;
-
-    let status = response.status();
-    let body = response.text().await?;
-    if !status.is_success() {
-        return Err(Error::Communication(body));
-    }
+impl QueryExecutor {
+    pub async fn create<Q: Into<QueryRequest>>(
```

Comment:
Could we implement `SnowflakeSession::execute` or a similar method to create a new query executor instead of calling `QueryExecutor::create` directly?

In the following examples, the second one appears to be more user-friendly than the first:

```rust
let session = ...;
let request = ...;
let executor = QueryExecutor::create(session, request);
```

```rust
let session = ...;
let request = ...;
let executor = session.execute(request);
```

We could retain this function for internal use.

Reply:
100% agree